File: contra/frontend/Source.java
package contra.frontend;

import java.io.BufferedReader;
import java.io.IOException;

/**
 * Source
 *
 * The framework class that represents the source program.
 *
 */
public class Source{
  public static final char EOL = '\n';      // End-Of-Line character
  public static final char EOF = (char) 0;  // End-Of-File character

  private BufferedReader reader;  // reader for the source program
  private String currentline;     // the contents of the current line
  private int currentlineNum;     // the current line number
  private int currentPos;         // the current postion in the current line

  /**
   * @Constructor
   *
   * @param  reader       The buffered reader for the source program
   * @throws IOException
   */
  public Source(BufferedReader reader) throws IOException{
    this.reader  = reader;
    this.currrentlineNum = 0;
    this.currentPos = -2;   // -2 indicates that the current line needs to be read in
  }

  /**
   * Return the source character at the current position.
   *
   * @return  The source character at the current position
   * @throws  Exception
   */
  public char currentChar() throws Exception {
    // First read of the source program?
    if(currentPos == -2) {
      // read in the first line and extract the first character
      readLine();
      return nextChar();
    }
    // End of the source program?
    else if(currentLine == null) {
      return EOF;
    }
    // End of the current line?
    else if((currentPos == -1) || (currentPos == currentLine.length())) {
      return EOL;
    }
    // Need to read the next line
    else if(currentPos > currentLine.length()) {
      readline();
      return nextChar();
    }
    // return the character at the current position
    else {
      return line.charAt(currentPos);
    }
  }

  /**
   * Consume the current character and return the next source character
   *
   * @return [description]
   */
  public char nextChar() throws Exception {
    ++currentPos;
    return currentChar();
  }

  /**
   * Return the source character following the current character
   * without consuming the current character.
   *
   * @return  The following character
   * @throws  Exception
   */
  public char peekChar() throws Exception {
    currentChar();
    if(line == null) {
      return EOF;
    }
  }

  /**
   * Read the next source line
   *
   * @throws IOException
   */
  private void readLine() throws IOException {
    line = reader.readLine();   // null at eof
    currentPos = -1;

    if(line != null) {
      ++lineNum;
    }
  }

  /**
   * Close the source
   *
   * @throws IOException
   */
   private void close() throws Exception {
     if(reader != null) {
       try {
         reader.close();
       } catch (IOException e){
         e.printStackTrace();
         throw ex;
       }
     }
   }
}

File: contra/frontend/Source.cpp
//
// Created by austindevx on 1/16/17.
//
#include <fstream>
#include "Source.h"

namespace frontend {
  char const Source::S_EOL = '\n';
  const char Source::S_EOF = EOF;

  Source::Source(std::string src_path) {
    stream.open(src_path, std::ios::in);
    currentLineNumber = 0;
    currentPosition = START_OF_FILE;

    // read in the first line to get things started
    readLine();
  }

  Source::~Source() { close(); }

  /**
   * Return the character at the current position in the source
   *
   * @return The character at the current position
  */
  char Source::getCurrentCharacter() {
    // if called before we start reading the current line then return the first character in the line
    if(currentPosition == START_OF_LINE) {
      return getNextCharacter();
    } else if(currentPosition == currentLine.length()) {
      // if stream.eof is true then we have reached the end of the input stream on the last read, when the characters
      // in the currentLine buffer are exhausted then return S_EOF to indicate the file stream has been exhausted as well
      // else we have just read the full line and the S_EOL before the next line is returned
      return stream.eof() ? S_EOF : S_EOL;
    } else if(currentPosition > currentLine.length()) {
      if(!stream.eof()) {
        // we have advanced past the end of the line so read in the next one
        readLine();
        return getNextCharacter();
      } else {
        return S_EOF;
      }
    } else {
      return currentLine.at(currentPosition);
    }
  }

  /**
   * Return the next character in the source
   *
   * @return The next character in the source
   */
  char Source::getNextCharacter() {
    currentPosition++;
    return getCurrentCharacter();
  }

  /**
   * Return the character at the next position in the source, without consuming
   * the current character
   *
   * @return The next character
   */
  char Source::peekNextCharacter() {
    if(char c = getCurrentCharacter() == S_EOF) {
      return S_EOF;
    }

    int nextPosition = currentPosition + 1;
    return nextPosition < currentLine.length() ? currentLine.at(nextPosition) : S_EOL;
  }

  void Source::readLine() {
    if(!stream.eof()) {
      std::getline(stream, currentLine);
      currentPosition = START_OF_LINE;
      currentLineNumber++;
    }
  }

  int Source::getCurrentLineNumber() {
    return currentLineNumber;
  }

  int Source::getCurrentPosition() {
    return currentPosition;
  }

  void Source::close() {
    if(stream.is_open()) stream.close();
  }
}


